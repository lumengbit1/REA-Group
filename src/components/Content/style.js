import styled from 'styled-components';

export const Root = styled.div`
  max-width: 1200px;
  display: flex;
  flex-direction: column;
  align-items: center;
`;

export const SubmitBtn = styled.button`
  width: 200px;
  height: 50px;
  margin: 20px 0;

  font-size: 20px;
`;

export const SelectContainer = styled.div`
  width: 100%;
`;
