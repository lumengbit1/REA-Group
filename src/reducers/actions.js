import axios from 'axios';
import { createAction } from 'redux-actions';
import queryString from 'query-string';
import _ from 'lodash';
import settings from '../settings';

export const get_location = createAction('GET_LOCATION_REQUEST');

export const weather_loading = createAction('WEATHER_LOADING', (loading) => loading);

export const location_loading = createAction('LOCATION_LOADING', (loading) => loading);

export const get_weather = createAction('GET_WEATHER_REQUEST');

export const get_location_successed = createAction('GET_LOCATION_RESOLVED');

export const get_weather_successed = createAction('GET_WEATHER_RESOLVED');

export const clear_data = createAction('CLEAR_DATA');

export const get_failed = createAction('GET_REJECTED');

export const getLocationAction = (location, params) => (dispatch) => {
  dispatch(get_location());
  dispatch(location_loading(true));

  return axios.get(`${settings.LOCATIONIQ_BASE_API_DOMAIN}?${queryString.stringify(_.assign({ q: location, key: settings.LOCATIONIQ_KEY, format: 'json' }, params))}`)
    .then((response) => {
      dispatch(location_loading(false));
      return dispatch(get_location_successed(response));
    })
    .catch((error) => dispatch(get_failed(error)));
};

export const getWeatherAction = (latitude, longitude, params) => (dispatch) => {
  dispatch(get_weather());
  dispatch(weather_loading(true));

  return axios.get(`${settings.OPEN_WEATHER_MAP_BASE_API_DOMAIN}?${queryString.stringify(_.assign({ lat: latitude, lon: longitude, appid: settings.OPEN_WEATHER_MAP_KEY, units: 'metric', exclude: 'current,minutely,hourly' }, params))}`)
    .then((response) => {
      dispatch(weather_loading(false));
      return dispatch(get_weather_successed(response));
    })
    .catch((error) => dispatch(get_failed(error)));
};
