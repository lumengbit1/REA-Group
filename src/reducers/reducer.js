import { handleActions } from 'redux-actions';
import { fromJS, List, Map } from 'immutable';
import {
  get_location_successed,
  get_weather_successed,
  get_failed,
  clear_data,
  get_location,
  get_weather,
  weather_loading,
  location_loading,
} from './actions';

const initialState = fromJS({
  location: [],
  weather: {},
  errors: {},
  loading: {
    weather: undefined,
    location: undefined,
  },
});

const reducer = handleActions(
  {
    [get_location]: (state) => state,
    [get_weather]: (state) => state,
    [weather_loading]: (state, action) => {
      const records = fromJS(action.payload);
      return state.setIn(['loading', 'weather'], records);
    },
    [location_loading]: (state, action) => {
      const records = fromJS(action.payload);
      return state.setIn(['loading', 'location'], records);
    },
    [get_location_successed]: (state, action) => {
      const records = fromJS(action.payload.data);
      return state.set('location', records);
    },
    [get_weather_successed]: (state, action) => {
      const records = fromJS(action.payload.data);
      return state.set('weather', records);
    },
    [get_failed]: (state, action) => {
      const errors = fromJS(action.payload.data);
      return state.set('errors', errors);
    },
    [clear_data]: (state) => state.set('location', List()).set('weather', Map()),
  },
  initialState,
);

export default reducer;
