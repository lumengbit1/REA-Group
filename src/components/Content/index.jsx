import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import { connect } from 'react-redux';
import { fromJS, Map } from 'immutable';

import Select from 'react-select';
import { getLocationAction, getWeatherAction, clear_data } from '../../reducers/actions';
import { makeLocationSelector, makeWeatherSelector, makeWeatherLoadingSelector, makeLocationLoadingSelector } from '../../selectors/selectors';
import WeatherWidget from '../WeatherWidget';
import { Root, SubmitBtn, SelectContainer } from './style';

function usePrevious(value) {
  const ref = React.useRef();
  React.useEffect(() => {
    ref.current = value;
  });
  return ref.current;
}

const Content = (props) => {
  const { getLocation, locationRecords, getWeatherReport, weatherRecords, ClearData, weatherLoading, locationLoading } = props;
  const [inputValue, setInputValue] = React.useState();

  const prevInputValue = usePrevious({ inputValue });

  const isActive = inputValue && inputValue.get('value').get('lat') && inputValue.get('value').get('lon');
  const latitude = inputValue && inputValue.get('value').get('lat');
  const longitude = inputValue && inputValue.get('value').get('lon');

  const options = locationRecords && locationRecords.size > 0
    ? locationRecords.map((record) => _.assign({}, { label: record.get('display_name'), value: { lat: record.get('lat'), lon: record.get('lon') } })).toJS()
    : [];

  const handleInputChange = _.debounce((value) => {
    if (value) {
      getLocation(value);
    }
  }, 500);

  const handleOnChange = (value) => {
    setInputValue(fromJS(value));
  };

  const handleOnSubmit = () => {
    getWeatherReport(latitude, longitude);
  };

  React.useEffect(() => {
    if (prevInputValue && prevInputValue.inputValue && !inputValue) {
      ClearData();
    }
  }, [inputValue]);

  return (
    <Root>
      <SelectContainer>
        <Select
          value={Map(inputValue).toJS()}
          onInputChange={handleInputChange}
          onChange={handleOnChange}
          options={options}
          isClearable
          isLoading={locationLoading}
        />
      </SelectContainer>
      <SubmitBtn
        disabled={!isActive}
        onClick={handleOnSubmit}
      >
        Submit
      </SubmitBtn>

      <WeatherWidget
        weatherRecords={weatherRecords}
        loading={weatherLoading}
      />

    </Root>
  );
};

Content.propTypes = {
  getLocation: PropTypes.func.isRequired,
  getWeatherReport: PropTypes.func.isRequired,
  ClearData: PropTypes.func.isRequired,
  locationRecords: ImmutablePropTypes.list,
  weatherRecords: ImmutablePropTypes.map,
  weatherLoading: PropTypes.bool,
  locationLoading: PropTypes.bool,
};

Content.defaultProps = {
  locationRecords: undefined,
  weatherRecords: undefined,
  weatherLoading: undefined,
  locationLoading: undefined,
};

export const makeMapStateToProps = () => {
  const getLocationSelector = makeLocationSelector();
  const getWeatherSelector = makeWeatherSelector();
  const getWeatherLoadingSelector = makeWeatherLoadingSelector();
  const getLocationLoadingSelector = makeLocationLoadingSelector();

  const mapStateToProps = (state) => ({
    locationRecords: getLocationSelector(state),
    weatherRecords: getWeatherSelector(state),
    weatherLoading: getWeatherLoadingSelector(state),
    locationLoading: getLocationLoadingSelector(state),
  });
  return mapStateToProps;
};

export const mapDispatchToProps = (dispatch) => ({
  getLocation: (location) => dispatch(getLocationAction(location)),
  getWeatherReport: (lat, lon) => dispatch(getWeatherAction(lat, lon)),
  ClearData: () => dispatch(clear_data()),
});

export default connect(
  makeMapStateToProps,
  mapDispatchToProps,
)(Content);
