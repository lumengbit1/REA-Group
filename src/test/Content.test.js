import React from 'react';
import 'jest-styled-components';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Content, { mapDispatchToProps } from '../components/Content';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { fromJS } from 'immutable';
import sinon from 'sinon';
import { getLocationAction, getWeatherAction, clear_data } from "../reducers/actions";
import thunk from 'redux-thunk';

Enzyme.configure({ adapter: new Adapter() });

function setup() {
  const props = {
    handleOnSubmit: jest.fn(),
  }

  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);
  let store = mockStore();
  const enzymeWrapper = mount(
    <Provider store={store} >
      <Content {...props} />
    </Provider>)

  return {
    props,
    enzymeWrapper
  }
}

describe('mapStateToProps,mapDispatchToProps Testing', () => {
  let wrapper, store;
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);
  beforeEach(() => {
    const initialState = fromJS({
      location: [],
      weather: {},
      errors: {},
    });

    store = mockStore(initialState);
    wrapper = mount(
      <Provider store={store} >
        <Content />
      </Provider>)
  })

  it('1.getLocation Testing', () => {
    const dispatchSpy = sinon.spy();
    const { getLocation } = mapDispatchToProps(dispatchSpy);
    getLocation();
    const expectedAction = getLocationAction();
    const spyLastCall = dispatchSpy.args[0][0];
    expect(spyLastCall.types).toEqual(expectedAction.types);
  })

  it('2.getWeatherReport Testing', () => {
    const dispatchSpy = sinon.spy();
    const { getWeatherReport } = mapDispatchToProps(dispatchSpy);
    getWeatherReport();
    const expectedAction = getWeatherAction();
    const spyLastCall = dispatchSpy.args[0][0];
    expect(spyLastCall.types).toEqual(expectedAction.types);
  })

  it('3.ClearData Testing', () => {
    const dispatchSpy = sinon.spy();
    const { ClearData } = mapDispatchToProps(dispatchSpy);
    ClearData();
    const expectedAction = clear_data();
    const spyLastCall = dispatchSpy.args[0][0];
    expect(spyLastCall.types).toEqual(expectedAction.types);
  })
})

describe('render snapshop testing', () => {
  it('renders correctly', () => {
    let wrapper, store;
    const middlewares = [thunk];
    const mockStore = configureMockStore(middlewares);
    store = mockStore();
    wrapper = shallow(
      <Provider store={store} >
        <Content />
      </Provider>);

    expect(wrapper).toMatchSnapshot();
  });
})
