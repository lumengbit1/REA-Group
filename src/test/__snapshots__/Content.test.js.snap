// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render snapshop testing renders correctly 1`] = `ShallowWrapper {}`;
